<?xml version="1.0"?>
<sdf version="1.4">
  <model name="r2d2">
    <!--<robot name="r2d2">-->

      <link name="base_link">
        <visual name="visual">
          <geometry>
            <cylinder radius="${width}" length="${bodylen}"/>
          </geometry>
          <material name="blue"/>
        </visual>
        <collision name="collision">
          <geometry>
            <cylinder radius="${width}" length="${bodylen}"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="10" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <link name="right_leg">
        <visual name="visual">
          <geometry>
            <box size="${leglen} 0.1 0.2"/>
          </geometry>
          <origin xyz="0 0 -${leglen/2}" rpy="0 ${pi/2} 0"/>
          <material name="white"/>
        </visual>
        <collision name="collision">
          <geometry>
            <box size="${leglen} 0.1 0.2"/>
          </geometry>
          <origin xyz="0 0 -${leglen/2}" rpy="0 ${pi/2} 0"/>
        </collision>
        <inertial>
          <mass value="10" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="base_to_right_leg" type="fixed">
        <parent link="base_link"/>
        <child link="right_leg"/>
        <origin xyz="0 ${-1*(width+.02)} 0.25" />
      </joint>

      <link name="right_base">
        <visual name="visual">
          <geometry>
            <box size="${baselen} 0.1 0.1"/>
          </geometry>
          <material name="white"/>
        </visual>
        <collision name="collision">
          <geometry>
            <box size="${baselen} 0.1 0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="10" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="right_base_joint" type="fixed">
        <parent link="right_leg"/>
        <child link="right_base"/>
        <origin xyz="0 0 ${-leglen}" />
      </joint>

      <link name="right_front_wheel">
        <visual name="visual">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
          <material name="black"/>
        </visual>
        <collision name="collision">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <joint name="right_front_wheel_joint" type="continuous">
        <axis xyz="0 1 0" rpy="0 0 0" />
        <parent link="right_base"/>
        <child link="right_front_wheel"/>
        <origin xyz="${baselen*1/3} 0 -${wheeldiam/2+.05}" rpy="0 0 0"/>
      </joint>

      <link name="right_back_wheel">
        <visual name="visual">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
          <material name="black"/>
        </visual>
        <collision name="collision">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <joint name="right_back_wheel_joint" type="continuous">
        <axis xyz="0 1 0" rpy="0 0 0" />
        <parent link="right_base"/>
        <child link="right_back_wheel"/>
        <origin xyz="${baselen*-1/3} 0 -${wheeldiam/2+.05}" rpy="0 0 0"/>
      </joint>
      <link name="left_leg">
        <visual name="visual">
          <geometry>
            <box size="${leglen} 0.1 0.2"/>
          </geometry>
          <origin xyz="0 0 -${leglen/2}" rpy="0 ${pi/2} 0"/>
          <material name="white"/>
        </visual>
        <collision name="collision">
          <geometry>
            <box size="${leglen} 0.1 0.2"/>
          </geometry>
          <origin xyz="0 0 -${leglen/2}" rpy="0 ${pi/2} 0"/>
        </collision>
        <inertial>
          <mass value="10" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="base_to_left_leg" type="fixed">
        <parent link="base_link"/>
        <child link="left_leg"/>
        <origin xyz="0 ${1*(width+.02)} 0.25" />
      </joint>

      <link name="left_base">
        <visual name="visual">
          <geometry>
            <box size="${baselen} 0.1 0.1"/>
          </geometry>
          <material name="white"/>
        </visual>
        <collision name="collision">
          <geometry>
            <box size="${baselen} 0.1 0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="10" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="left_base_joint" type="fixed">
        <parent link="left_leg"/>
        <child link="left_base"/>
        <origin xyz="0 0 ${-leglen}" />
      </joint>
      <link name="left_front_wheel">
        <visual name="visual">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
          <material name="black"/>
        </visual>
        <collision name="collision">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <joint name="left_front_wheel_joint" type="continuous">
        <axis xyz="0 1 0" rpy="0 0 0" />
        <parent link="left_base"/>
        <child link="left_front_wheel"/>
        <origin xyz="${baselen*1/3} 0 -${wheeldiam/2+.05}" rpy="0 0 0"/>
      </joint>
      <link name="left_back_wheel">
        <visual name="visual">
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
          <material name="black"/>
        </visual>
        <collision>
          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
          <geometry>
            <cylinder radius="${wheeldiam/2}" length="0.1"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <joint name="left_back_wheel_joint" type="continuous">
        <axis xyz="0 1 0" rpy="0 0 0" />
        <parent link="left_base"/>
        <child link="left_back_wheel"/>
        <origin xyz="${baselen*-1/3} 0 -${wheeldiam/2+.05}" rpy="0 0 0"/>
      </joint>

      <joint name="gripper_extension" type="prismatic">
        <parent link="base_link"/>
        <child link="gripper_pole"/>
        <limit effort="1000.0" lower="-${width*2-.02}" upper="0" velocity="0.5"/>
        <origin rpy="0 0 0" xyz="${width-.01} 0 0.2"/>
        <limit effort="30" velocity="0.2"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>

      <link name="gripper_pole">
        <visual name="visual">
          <geometry>
            <cylinder length="${polelen}" radius="0.01"/>
          </geometry>
          <origin xyz="${polelen/2} 0 0" rpy="0 ${pi/2} 0 "/>
        </visual>
        <collision>
          <geometry>
            <cylinder length="${polelen}" radius="0.01"/>
          </geometry>
          <origin xyz="${polelen/2} 0 0" rpy="0 ${pi/2} 0 "/>
        </collision>
        <inertial>
          <mass value="0.05" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <!--<transmission name="gripper_extension_trans" type="SimpleTransmission">-->
        <!--<type>transmission_interface/SimpleTransmission</type>-->
        <!--<actuator name="gripper_extension_motor">-->
          <!--<mechanicalReduction>1</mechanicalReduction>-->
          <!--</actuator>-->
        <!--<joint name="gripper_extension">-->
          <!--<hardwareInterface>EffortJointInterface</hardwareInterface>-->
          <!--</joint>-->
        <!--</transmission>-->

      <joint name="left_gripper_joint" type="revolute">
        <axis xyz="0 0 1"/>
        <limit effort="1000.0" lower="0.0" upper="0.548" velocity="0.5"/>
        <origin rpy="0 0 0" xyz="${polelen} ${1*0.01} 0"/>
        <parent link="gripper_pole"/>
        <child link="left_gripper"/>
        <limit effort="30" velocity="1.0"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="left_gripper">
        <visual name="visual">
          <origin rpy="${(1-1)/2*pi} 0 0" xyz="0 0 0"/>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger.dae"/>-->
          </geometry>
        </visual>
        <collision>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger.dae"/>-->
          </geometry>
          <origin rpy="${(1-1)/2*pi} 0 0" xyz="0 0 0"/>
        </collision>
        <inertial>
          <mass value="0.05" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <!--<transmission name="left_gripper_trans" type="simpletransmission">-->
        <!--<type>transmission_interface/simpletransmission</type>-->
        <!--<actuator name="left_gripper_motor">-->
          <!--<mechanicalReduction>1</mechanicalReduction>-->
          <!--[><hardwareinterface>hardwareinterface/velocityjointinterface</hardwareinterface><]-->
          <!--</actuator>-->
        <!--<joint name="left_gripper_joint">-->
          <!--<hardwareInterface>PositionJointInterface</hardwareInterface>-->
          <!--</joint>-->
        <!--</transmission>-->

      <joint name="left_tip_joint" type="fixed">
        <parent link="left_gripper"/>
        <child link="left_tip"/>
      </joint>
      <link name="left_tip">
        <visual name="visual">
          <origin rpy="${(1-1)/2*pi} 0 0" xyz="0.09137 0.00495 0"/>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger_tip.dae"/>-->
          </geometry>
        </visual>
        <collision>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger_tip.dae"/>-->
          </geometry>
          <origin rpy="${(1-1)/2*pi} 0 0" xyz="0.09137 0.00495 0"/>
        </collision>
        <inertial>
          <mass value="0.05" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="right_gripper_joint" type="revolute">
        <axis xyz="0 0 -1"/>
        <limit effort="1000.0" lower="0.0" upper="0.548" velocity="0.5"/>
        <origin rpy="0 0 0" xyz="${polelen} ${-1*0.01} 0"/>
        <parent link="gripper_pole"/>
        <child link="right_gripper"/>
        <limit effort="30" velocity="1.0"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>
      <link name="right_gripper">
        <visual name="visual">
          <origin rpy="${(-1-1)/2*pi} 0 0" xyz="0 0 0"/>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger.dae"/>-->
          </geometry>
        </visual>
        <collision>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger.dae"/>-->
          </geometry>
          <origin rpy="${(-1-1)/2*pi} 0 0" xyz="0 0 0"/>
        </collision>
        <inertial>
          <mass value="0.05" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>
      <!--<transmission name="right_gripper_trans" type="SimpleTransmission">-->
        <!--<type>transmission_interface/SimpleTransmission</type>-->
        <!--<actuator name="right_gripper_motor">-->
          <!--<mechanicalReduction>1</mechanicalReduction>-->
          <!--[><hardwareInterface>hardwareInterface/VelocityJointInterface</hardwareInterface><]-->
          <!--</actuator>-->
        <!--<joint name="right_gripper_joint">-->
          <!--<hardwareInterface>PositionJointInterface</hardwareInterface>-->
          <!--</joint>-->
        <!--</transmission>-->

      <joint name="right_tip_joint" type="fixed">
        <parent link="right_gripper"/>
        <child link="right_tip"/>
      </joint>
      <link name="right_tip">
        <visual name="visual">
          <origin rpy="${(-1-1)/2*pi} 0 0" xyz="0.09137 0.00495 0"/>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger_tip.dae"/>-->
          </geometry>
        </visual>
        <collision>
          <geometry>
            <!--<mesh filename="package://r2d2/meshes/l_finger_tip.dae"/>-->
          </geometry>
          <origin rpy="${(-1-1)/2*pi} 0 0" xyz="0.09137 0.00495 0"/>
        </collision>
        <inertial>
          <mass value="0.05" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <link name="head">
        <visual name="visual">
          <geometry>
            <sphere radius="${width}"/>
          </geometry>
          <material name="white"/>
        </visual>
        <collision>
          <geometry>
            <sphere radius="${width}"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="2" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="head_swivel" type="continuous">
        <parent link="base_link"/>
        <child link="head"/>
        <axis xyz="0 0 1"/>
        <origin xyz="0 0 ${bodylen/2}"/>
        <limit effort="30" velocity="1.0"/>
        <dynamics damping="0.0" friction="0.0"/>
      </joint>

      <!-- This block connects the head_swivel joint to an actuator (motor), which informs both
      simulation and visualization of the robot -->
      <!--<transmission name="head_swivel_trans">-->
        <!--<type>transmission_interface/SimpleTransmission</type>-->
        <!--<actuator name="$head_swivel_motor">-->
          <!--<mechanicalReduction>1</mechanicalReduction>-->
          <!--</actuator>-->
        <!--<joint name="head_swivel">-->
          <!--<hardwareInterface>PositionJointInterface</hardwareInterface>-->
          <!--</joint>-->
        <!--</transmission>-->

      <link name="box">
        <visual name="visual">
          <geometry>
            <box size="0.08 0.08 0.08"/>
          </geometry>
          <material name="blue"/>
          <origin xyz="-0.04 0 0"/>
        </visual>
        <collision>
          <geometry>
            <box size="0.08 0.08 0.08"/>
          </geometry>
        </collision>
        <inertial>
          <mass value="1" />
          <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
      </link>

      <joint name="tobox" type="fixed">
        <parent link="head"/>
        <child link="box"/>
        <origin xyz="${.707*width+0.04} 0 ${.707*width}"/>
      </joint>

      <!-- Gazebo plugin for ROS Control -->
      <!--<gazebo>-->
        <!--<plugin name="r2d2_arm" filename="libr2d2_arm.so"/>-->
        <!--<modelNamespace>/</modelNamespace>-->
        <!--</plugin>-->
      <!--</gazebo>-->

  <!--</robot>-->
  </model>
</sdf>
